Add transformation from and to the `ghc-lib` AST data structure

stu204767 opened the pull request:
### Issue
Closes #33.

### Description of the Change
Adds a two-way transformation between the AST data structure used by `ghc-lib` and our own AST data structure.

### Verification Process
The transformation has been tested in #43.

### Additional Notes
There are some constructs (for example statements in guards) where I have tested how they are represented in the `ghc-lib` AST for basic cases, but where I'm not sure if this is the only possible representation that we can support. I've marked these places with `TODO` comments in the code.

stu204767 committed:
Add "header" of `FromGHC` module and `ghc-lib-parser` dependencies #33

stu204767 committed:
Replace `DeclHead` type with `Name` #33

stu204767 committed:
Implement transformation of GHC data types to HST data types #33

stu204767 committed:
Add remaining transformation functions to `FromGHC.hs` #33

stu204767 committed:
Add `EqAST` and `ShowAST` instances to GHC type instances #33

stu204767 committed:
Merge branch 'main' into issue-33

stu204767 committed:
Add first part of transformation functions to `ToGHC.hs` #33

stu204767 committed:
Merge branch 'main' into issue-33 and resolve merge conflicts

stu204767 committed:
Apply HLint suggestions #33

stu204767 committed:
Simplify `ConDecl` type in `Syntax.hs` (doesn't store the types anymore, only the number of them) #33

stu204767 committed:
Simplify `TypeWrapper` in `FromGHC.hs` (types in constructor declarations are no longer needed) #33

stu204767 committed:
Add remaining transformation functions to `ToGHC.hs` #33

stu204767 committed:
Simplify the transformation of function names in matches #33

stu204767 committed:
Apply HLint suggestion #33

stu204767 committed:
Improve TODO and notes commments #33

stu204767 committed:
Add Haddock comments to `FromGHC.hs` #33

stu204767 committed:
Add Haddock comments to `ToGHC.hs` #33

stu204767 committed:
Rearrange some functions #33

stu204767 committed:
Merge branch 'main' into issue-33

stu204767 committed:
Adjust code to the merged changes from the main branch #33

just95 commented on haskell-src-transformations.cabal:
```diff
@@ -31,14 +31,16 @@ flag PolysemyPluginEnabled {
+                     , containers        >= 0.6       && < 0.7
+                     , directory         >= 1.3       && < 1.4
+                     , extra             >= 1.7       && < 1.8
+                     , ghc-lib-parser    >= 8.6       && < 9.0
+                     , ghc-lib-parser-ex >= 8.10.0.15 && < 9.0
```
The version constraint `8.10.0.15` is oddly specific. Doesn't `8.10` suffice? Is backward compatibility of all versions up to `9.0` guaranteed by `ghc-lib`'s versioning scheme? If not, the upper bounds should be `8.7` and `8.11`, respectively.

stu204767 replied:
The reason for this long version number is that `8.10` is the GHC version that is being built upon. `ghc-lib-parser-ex` is a fairly new package and has been restructured quite a bit under the GHC version `8.10`, which is why we have to be this specific. The upper bound is therefore very optimistic and should propably be `8.10.0.16` to be sure.

just95 commented on haskell-src-transformations.cabal:
```diff
@@ -84,6 +87,8 @@ library
                     , HST.Frontend.FromHSE
                     , HST.Frontend.ToHSE
+                    , HST.Frontend.FromGHC
+                    , HST.Frontend.ToGHC
```
```suggestion
                    , HST.Frontend.FromGHC
                    , HST.Frontend.FromHSE
                    , HST.Frontend.Syntax
                    , HST.Frontend.ToGHC
                    , HST.Frontend.ToHSE
```

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,382 @@
+--   used by @ghc-lib-parser@.
+data LitWrapper = Lit (GHC.HsLit GHC.GhcPs)
+                | OverLit (GHC.HsOverLit GHC.GhcPs)
+  deriving Eq
```
Shouldn't literals also be compared with `defaultPrintEq` such that source spans are ignored?

stu204767 replied:
These literal types don't appear to contain source span information.

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,382 @@
+  , (GHC.listTyConName                    , S.ListCon s)
+  , (GHC.tyConName GHC.funTyCon           , S.FunCon s)
+  , (GHC.consDataConName                  , S.Cons s)
+  , (GHC.tupleTyConName GHC.UnboxedTuple 1, S.UnboxedSingleCon s)
```
`S.UnboxedSingleCon` is `(# #)` but `GHC.tupleTyConName GHC.UnboxedTuple 1` sounds like `(# τ #)` to me.

stu204767 replied:
Okay, I thought `SingleCon` meant one element, but I will change that to a `0` then.
All of these special constructors need to be tested once we start testing with `ghc-lib` (I wasn't able to do this yet, because most AST components of `ghc-lib-parser` don't have `Show` instances and the `showAstData` function simplifies the representation of names).

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,382 @@
+-- | Takes an HST source span and maps GHC names to HST special constructors.
+specialConMap :: S.SrcSpan GHC -> [(GHC.Name, S.SpecialCon GHC)]
+specialConMap s =
+  [ (GHC.tyConName GHC.unitTyCon          , S.UnitCon s)
+  , (GHC.listTyConName                    , S.ListCon s)
+  , (GHC.tyConName GHC.funTyCon           , S.FunCon s)
+  , (GHC.consDataConName                  , S.Cons s)
+  , (GHC.tupleTyConName GHC.UnboxedTuple 1, S.UnboxedSingleCon s)
+  ]
```
That the special case for `S.TupleCon` is not covered by `specialConMap` but is implemented in `transformName` should at least be documented. However, for error reporting purposes it would be better if all names were transformed in `transformName` using pattern matching.

stu204767 replied:
I wanted to use pattern matching, but the [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#v:Name) and [`OccName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/OccName.html#t:OccName) types of `ghc-lib-parser` don't export their constructors, which is why I chose this solution.
I will mention the special case for tuple constructors in the documentation, though.

stu204146-kiel commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -78,9 +78,9 @@ deriving instance ShowAST a => Show (Decl a)
-data ConDecl a = ConDecl (Name a) [TypeExp a]
-               | InfixConDecl (TypeExp a) (Name a) (TypeExp a)
-deriving instance EqAST a => Eq (ConDecl a)
+data ConDecl a = ConDecl (Name a) Int
```
Why did you remove the `TypeExp`s in `ConDecl`?

just95 replied:
The `TypeExp`s are not needed since `ConDecl`s are not converted back to the original AST. The only information that we need about constructor declarations is their arity.

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,382 @@
+  show (OverLit l) = defaultPrintShow l
+
+-- | Wrapper for the type for type expressions appearing in type signatures
+--   used by @ghc-lib-parser@.
+newtype TypeWrapper = SigType (GHC.LHsSigWcType GHC.GhcPs)
+
+instance Eq TypeWrapper where
+  SigType t1 == SigType t2 = defaultPrintEq t1 == defaultPrintEq t2
+
+instance Show TypeWrapper where
+  show (SigType t) = defaultPrintShow t
+
+-------------------------------------------------------------------------------
+-- Printing Functions for the @ghc-lib-parser@ AST                           --
+-------------------------------------------------------------------------------
+
+-- | Prints a component of the @ghc-lib-parser@ AST data structure without
+--   source span information and with line breaks.
+--
+--   This function is used for equality of @ghc-lib-parser@ AST components.
+defaultPrintEq :: Data a => a -> String
+defaultPrintEq d =
+  GHC.showSDoc defaultDynFlags (GHC.showAstData GHC.BlankSrcSpan d)
+
+-- | Prints a component of the @ghc-lib-parser@ AST data structure with source
+--   span information and without line breaks.
+--
+--   This function is used for showing @ghc-lib-parser@ AST components.
+defaultPrintShow :: Data a => a -> String
+defaultPrintShow d =
+  GHC.showSDocOneLine defaultDynFlags (GHC.showAstData GHC.NoBlankSrcSpan d)
+
+-- | Returns default variants of the dynamic flags used by @ghc-lib-parser@.
+--
+--   These dynamic flags are used for parsing Haskell code and printing AST
+--   components.
+defaultDynFlags :: GHC.DynFlags
+-- TODO These DynFlags should be the same as the ones used for parsing Haskell
+-- modules and should be defined in another module.
+defaultDynFlags = GHC.defaultDynFlags GHC.fakeSettings GHC.fakeLlvmConfig
```
Since the type family instances start to become more complicated, I'd suggest moving them to their own modules (e.g. `HST.Frontend.Syntax.GHC` and `HST.Frontend.Syntax.HSE`)

stu204767 replied:
I would call these modules `HST.Frontend.GHC` and `HST.Frontend.HSE`, so without `Syntax`, because they are not a part of our syntax.
We could also create subdirectories `GHC` and `HSE` in the `Frontend` directory and call these modules `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.HSEConfig` or similar.

just95 replied:
I considered the `type instance`s part of the syntax since the `type family` declarations are in the `Syntax` module, too. However, `GHC` and `HSE` subdirectories are fine as well. `From*` and `To*` could be moved to the subdirectories as well. I'd just avoid using `GHC` and `HSE` in module names twice, i.e., `HST.Frontend.GHC.Config` and `HST.Frontend.GHC.From` instead of `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.FromGHC`.

just95 commented on src/lib/HST/Frontend/Syntax.hs:
```diff
@@ -78,9 +78,9 @@ deriving instance ShowAST a => Show (Decl a)
-data ConDecl a = ConDecl (Name a) [TypeExp a]
-               | InfixConDecl (TypeExp a) (Name a) (TypeExp a)
-deriving instance EqAST a => Eq (ConDecl a)
+data ConDecl a = ConDecl (Name a) Int
+               | InfixConDecl (Name a)
```
Only the `ConDecl` constructor and a `isInfix` flag would probably suffice, too.

stu204767 replied:
Do you want a new type for this flag or a `Bool`?
Either way, I don't think two constructors for `ConDecl` or a flag make much of a difference, but I will change it if you want.

just95 replied:
> I don't think two constructors for `ConDecl` or a flag make much of a difference

I agree. The only difference it would make is that we could reduce some code duplication in `HST.Application`.

> Do you want a new type for this flag or a Bool?

I'd suggest something like this.

```suggestion
data ConDecl a
  = ConDecl { conDeclName    :: Name a
            , conDeclArity   :: Int
            , conDeclIsInfix :: Bool
            }
```

stu204767 committed:
Apply review suggestions #33

just95 committed:
Report parse errors #33 #38

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,395 @@
+      , GHC.isDataOcc name
+      )
+transformRdrName (GHC.L s (GHC.Exact name)) =
+  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), False)
```
I was testing with the queue `./example`s and got the following error message whenever the module contained patterns for built-in data types.

> ```
> Internal: Only constructors can be applied in patterns
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:324:14 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

As far as I can tell, the problem is that this function returns `False` for special constructors. Do you think that it is safe to change this to `True`?

```suggestion
  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), True)
```

stu204767 replied:
Yes, that should be safe. I didn't consider that these built-in types could appear in patterns as well.

just95 replied:
See a69a8cf230c47c4c20293513a67c42e280f05ffd.

just95 commented on src/lib/HST/Frontend/FromGHC.hs:
```diff
@@ -0,0 +1,395 @@
+-- | Transforms a GHC name with an HST source span into an HST special
+--   constructor.
+transformName :: S.SrcSpan GHC -> GHC.Name -> S.SpecialCon GHC
+transformName s name = case lookup name (specialConMap s) of
```
With the patch above, unit and empty list patterns still caused the following error message in my tests.

> ```
> Internal: Expected a built-in special constructor
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:376:10 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

However, pair patterns worked correctly which suggests that the `lookup` in `specialConMap` probably does not work.

stu204767 replied:
Okay, then we should probably not use the `Eq` instance of [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#t:Name) at all, it also has a note saying it should be used with caution.
What do you think about using `nameOccName :: Name -> OccName` and `occNameString :: OccName -> String` and implementing `specialConMap` on those strings?
Or, as this probably does retain the spelling of the built-in constructor, so `()` wouldn't be equal to `( )`, a pretty-printing function could be used instead.

just95 replied:
I did some more tests. As it turns out `ghc-lib` differentiates between names of type and data constructors. Unlike in `haskell-src-exts`, the list type constructor [`GHC.listTyConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:listTyConName) is not the same as the empty list constructor [`GHC.nilDataConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:nilDataConName), for example. Since we never need type constructors, I've changed `specialConMap` to use the data constructor names in 1af73e6c3df8cda6b987b2aa531f0254c244b52b. Furthermore, I had to change `GHC.ATyCon` to `GHC.AConLike` in `transformName` to support tuples in infix notation (e.g. `(,) x y`).

just95 committed:
Implement GHC-lib front end in `Main` #33

just95 committed:
Fix transformation of special constructors #33

See <https://github.com/FreeProving/haskell-src-transformations/pull/40#discussion_r461132596>.

just95 committed:
Fix transformation of tuple constructors #33

just95 committed:
Add type classes for parsing, transforming and pretty-printing #33

The type classes allow us to generalize processing of input
modules in `Main` and will make it easier to run tests for
both front end as well.

just95 committed:
Add effect for front end operations #33

just95 committed:
Use `WithFrontend` effect in `ApplicationTests` #33

just95 committed:
Add effect for setting Hspec expectations #33

just95 committed:
Remove unused language extensions #33

stu204767 committed:
Merge branch 'issue-38' into issue-33

stu204767 committed:
Adapt `FromGHC.hs` to merged changes in the syntax #33

stu204767 committed:
Adapt `ToGHC.hs` to merged changes in the syntax #33

stu204767 committed:
Restructure front ends #33

stu204767 committed:
Fix restructuring errors in tests #33

stu204767 committed:
Add and update Haddock and other comments #33

stu204767 committed:
Apply simple suggestions from my own code review

just95 committed:
Apply suggestions from code review #43

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Set line length to 80 columns in HSE pretty printer #43

just95 committed:
Rename destructors of `ParsedModule` #43

just95 committed:
Merge branch 'issue-33-tests' of github.com:FreeProving/haskell-src-transformations into issue-33-tests

just95 committed:
Merge branch 'issue-33' into issue-33-tests

just95 committed:
Remove input module from `unTransformModule` #33

just95 committed:
Remove `S.FunCon` #43

See <https://github.com/FreeProving/haskell-src-transformations/pull/43#discussion_r462461038>.

just95 committed:
Format code #43

stu204767 committed:
Switch to the report effect in `GHC/From.hs` #38

stu204767 committed:
Improve reported messages in `GHC/From.hs` #33
For this, pattern matching was expanded to explicitely include most unsupported constructors

stu204767 committed:
Improve comments #33

stu204767 committed:
Switch to the report effect in `GHC/To.hs` #38

stu204767 committed:
Move functions for AST transformation error messages to `Effect/Report.hs` #33

stu204767 committed:
Apply HLint suggestions #33

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+  let modName' = case GHC.hsmodName modul of
+        Just (GHC.L s' modName) ->
+          Just (transformModuleName (transformSrcSpan s') modName)
+        Nothing -> Nothing
```
If you change the type of `transformModuleName` from `S.SrcSpan GHC -> GHC.ModuleName -> S.ModuleName GHC` to `GHC.Located GHC.ModuleName -> S.ModuleName GHC` you can simply write the following here.

```suggestion
  let modName' = transformModuleName <$> GHC.hsmodName modul
```

just95 commented on src/lib/HST/Frontend/GHC/To.hs:
```diff
@@ -0,0 +1,452 @@
+transformSpecialCon (S.UnboxedSingleCon _) =
+  return $ GHC.tupleTyConName GHC.UnboxedTuple 0
+transformSpecialCon (S.ExprHole _) =
+  reportFatal
+    $  Message Error
+    $  "Encountered expression hole at name level in retransformation! "
+    ++ "(Expression holes should be transformed at expression level with "
+    ++ "the ghc-lib front end)"
```
```suggestion
  reportFatal
    $  Message Internal
    $  "Encountered expression hole at name level in retransformation. "
    ++ "Expression holes should be transformed at expression level with "
    ++ "the ghc-lib front end!"
```

stu204767 replied:
I suppose the other two `Message`s in this module should be `Internal` as well?

just95 replied:
Yes :+1:
An `Error` tells the user that they did something wrong whereas an `Internal` error points us to a bug.

just95 commented on src/lib/HST/Effect/Report.hs:
```diff
@@ -21,12 +21,17 @@ module HST.Effect.Report
+  , failToReport
+    -- * Error Messages for AST Transformation
+  , notSupported
+  , skipNotSupported
```
I don't think that it is a good idea to clutter this module with concrete error messages. Let's move them to their own module (e.g, `HST.Frontend.Common.Messages` or `HST.Frontend.Transformer.Messages`).

just95 committed:
Merge branch 'issue-33' into issue-33-tests

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+transformDataDefn GHC.HsDataDefn { GHC.dd_cons = cons } = do
+  conDecls <- mapM transformConDecl cons
+  return $ if all isJust conDecls then Just (catMaybes conDecls) else Nothing
+transformDataDefn (GHC.XHsDataDefn _) = return Nothing
```
From the [documentation of `XXHsDataDefn`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/src/GHC.Hs.Decls.html#HsDataDefn) it looks like the field of `XHsDataDefn` always has type `NoExtCon` (which has no constructors) if `GHC.GhcPs` is used. Thus, we could also write the following to statically ensure that we are not missing a constructor.

```suggestion
transformDataDefn (GHC.XHsDataDefn x) = noExtCon x
```

Otherwise, we should probably inform the user that the declaration is skipped. If no warning is printed, they might be surprised that constructors of some data types are not handled correctly.

```suggestion
transformDataDefn (GHC.XHsDataDefn _) = do
  skipNotSupported "Data type declaration extension"
  return Nothing
```

stu204767 replied:
`noExtCon` can be used at almost all of these `X`-constructors, so I've changed this at all of those places. Interestingly, the extension of [`HsValBindsLR`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/GHC-Hs-Binds.html#t:HsValBindsLR) remains as the only one of these extensions that doesn't equal to `NoExtCon`.

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+  :: Member Report r
+  => GHC.HsDataDefn GHC.GhcPs
+  -> Sem r (Maybe [S.ConDecl GHC])
+transformDataDefn GHC.HsDataDefn { GHC.dd_cons = cons } = do
+  conDecls <- mapM transformConDecl cons
+  return $ if all isJust conDecls then Just (catMaybes conDecls) else Nothing
```
The [documentation of `HsDataDefn`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/src/GHC.Hs.Decls.html#HsDataDefn) suggests that the constructor can represent data family instances, too, which are not supported by us.

stu204767 replied:
Yes, but those should not occur under the `TyClD` constructor of [`HsDecl`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/GHC-Hs-Decls.html#t:HsDecl), but under `InstD` -> `DataFamInstD` -> `DataFamInstDecl`. This type also contains an `HsDataDefn`.

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+  -> Sem r (S.Exp GHC)
+transformStmtExpr (GHC.L _ (GHC.BodyStmt _ body _ _)) = transformExpr body
+transformStmtExpr _ = notSupported "Complex guards"
+-- TODO Are there more statements that can be safely converted to boolean expressions?
```
I'd prefer to handle every constructor explicitly even if all of other constructors are `notSupported`.

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+transformExpr (GHC.L _ (GHC.XExpr _)) = notSupported "Expression extensions"
+
+
+-- | Transforms a GHC located tuple argument consisting of an expression into
```
```suggestion

-- | Transforms a GHC located tuple argument consisting of an expression into
```

just95 commented on src/lib/HST/Frontend/GHC/From.hs:
```diff
@@ -0,0 +1,605 @@
+      Just GHC.BoxedTuple -> return $ S.TupleCon s S.Boxed (GHC.tyConArity tc)
+      Just GHC.UnboxedTuple ->
+        return $ S.TupleCon s S.Unboxed (GHC.tyConArity tc)
+      _ -> notSupported "Unexpected special constructor"
+    _ -> notSupported "Expected a built-in special constructor"
```
`notSupported` doesn't really fit in this case. However, since this code is changed in #43 anyway, we can probably leave it as is for now.

stu204767 replied:
That's the reason I haven't changed these messages.

stu204767 committed:
Apply review suggestions #33

just95 committed:
Merge branch 'issue-33' into issue-33-tests

stu204767 committed:
Order HLint aliases #33

just95 committed:
Apply suggestions from code review #43

Co-authored-by: stu204767 <63296935+stu204767@users.noreply.github.com>

just95 committed:
Format code in test cases #43

just95 committed:
Remove superfluous `do` #43

stu204767 committed:
Merge branch 'main' into issue-33

just95 committed:
Merge pull request #43 from FreeProving/issue-33-tests

Integrate GHC-lib into `Main` module and unit tests

just95 committed:
Fix identation of Haddock comments #40

just95 committed:
Fix out of scope identifiers in Haddock #40
